lib/ttfunk/table/os2.rb

```diff
@@ -40,36 +42,246 @@ class OS2 < Table
+        (0xA880..0xA8DF) => 115,   (0xA900..0xA92F) => 116,
+        (0xA930..0xA95F) => 117,   (0xAA00..0xAA5F) => 118,
+        (0x10190..0x101CF) => 119, (0x101D0..0x101FF) => 120,
+        (0x102A0..0x102DF) => 121, (0x10280..0x1029F) => 121,
+        (0x10920..0x1093F) => 121, (0x1F030..0x1F09F) => 122,
+        (0x1F000..0x1F02F) => 122
+      }.freeze
+
+      UNICODE_MAX = 0xFFFF
+      UNICODE_RANGES = UNICODE_BLOCKS.keys.freeze
+
       def tag
         'OS/2'
       end
 
+      class << self
+        def encode(os2, subset)
+          ''.tap do |result|
+            result << [
+              os2.version, os2.ave_char_width, os2.weight_class,
+              os2.width_class, os2.type, os2.y_subscript_x_size,
+              os2.y_subscript_y_size, os2.y_subscript_x_offset,
+              os2.y_subscript_y_offset, os2.y_superscript_x_size,
+              os2.y_superscript_y_size, os2.y_superscript_x_offset,
+              os2.y_superscript_y_offset, os2.y_strikeout_size,
+              os2.y_strikeout_position, os2.family_class
+            ].pack('n*')
+
+            result << os2.panose
+
+            new_char_range = unicode_blocks_for(os2, os2.char_range, subset)
+            result << BinUtils
+                      .unpack_int(new_char_range.value, 32)
+                      .pack('N*').ljust(16, "\0")
+
+            result << os2.vendor_id
+
+            new_cmap_table = subset.new_cmap_table[:charmap]
+            code_points = new_cmap_table
+                          .keys
+                          .select { |k| new_cmap_table[k][:new] > 0 }
+                          .sort
+
+            # "This value depends on which character sets the font supports.
+            # This field cannot represent supplementary character values
+            # (codepoints greater than 0xFFFF). Fonts that support
+            # supplementary characters should set the value in this field
+            # to 0xFFFF."
+            first_char_index = [code_points.first || 0, UNICODE_MAX].min
+            last_char_index = [code_points.last || 0, UNICODE_MAX].min
+
+            result << [
+              os2.selection, first_char_index, last_char_index
+            ].pack('n*')
+
+            if os2.version > 0
+              result << [
+                os2.ascent, os2.descent, os2.line_gap,
+                os2.win_ascent, os2.win_descent
+              ].pack('n*')
+
+              result << BinUtils
+                        .unpack_int(code_pages_for(subset).value, 32)
+                        .pack('N*').ljust(8, "\0")
+
+              if os2.version > 1
+                result << [
+                  os2.x_height, os2.cap_height, os2.default_char,
+                  os2.break_char, os2.max_context
+                ].pack('n*')
+              end
+            end
+          end
+        end
+
+        private
+
+        def code_pages_for(subset)
+          field = BitField.new(0)
+          return field if subset.unicode?
```

Comment:
Hmm good question. After some additional research I was able to find several of the Noto fonts that have a `code_page_range` of 0 but don't include the majority of the 1252 code page (i.e. a-z, A-Z). So I guess there's precedent for it? Since TTFunk's subsets are very encoding-based, I don't know how we would do this any differently.

Reply:
[Spec](https://docs.microsoft.com/en-us/typography/opentype/spec/os2ver1#cpr) suggests only setting encoding bit if the code table is functional. However…

> The determination of “functional” is left up to the font designer

Also this bit:

> If the font file is encoding ID 0, then the Symbol Character Set bit should be set.

Is this handled?

Reply:
I believe an encoding ID of 0 only signifies a symbol character set on Microsoft platforms, i.e. platform ID 3. Currently TTFunk doesn't support that combination in the [cmap logic](https://github.com/camertron/ttfunk/blob/master/lib/ttfunk/table/cmap/subtable.rb#L13), so we should be ok.